mcdemoaux/vision/neural.py

```diff
@@ -8,35 +8,44 @@
 model_cache = dict()
 
 
-def process_pixel_data(pixels, resize, scale):
-    # img_data = numpy.frombuffer(pixels, dtype=numpy.uint8)
-    img_data = cv2.resize(pixels, dsize=(320 * scale, 240 * scale), interpolation=cv2.INTER_CUBIC)
-    # img_data = numpy.resize(pixels, (240 * scale, 320 * scale, 4))
-    # img_data = pixels
-    if resize != 1:
-        height, width, _ = img_data.shape
-        img_data = cv2.resize(img_data, (int(width * resize), int(height * resize)),
-            fx=resize, fy=resize, interpolation=cv2.INTER_NEAREST)
+def process_pixel_data(pixels, keep_aspect_ratio, maximum_size):
+    height, width, _ = pixels.shape
+    aspect_ratio_org = width / height
+    wscale = width // 320
+    scaled_width = int(320*wscale)
+    scaled_height = int(int(keep_aspect_ratio) * (scaled_width / aspect_ratio_org) + (1 - int(keep_aspect_ratio)) * int(240*wscale))
+    if maximum_size is not None and maximum_size:
+        scaled_width = maximum_size[0] if scaled_width > maximum_size[0] else scaled_width
+        if not keep_aspect_ratio:
+            scaled_height = maximum_size[1] if scaled_height > maximum_size[1] else scaled_height
+        else:
+            scaled_height = int(scaled_width / aspect_ratio_org)
+
+    # to make width and height divisible by 8
+    scaled_width -= scaled_width % 8
+    scaled_height -= scaled_height % 8
+
+    img_data = cv2.resize(pixels, dsize=(scaled_width, scaled_height), interpolation=cv2.INTER_CUBIC)
 
     return img_data
 
 
-def get_image(img_frame, resize, scale):
+def get_image(img_frame, keep_aspect_ratio, maximum_size):
     if img_frame is not None:
-        return process_pixel_data(img_frame.pixels, resize, scale)
+        return process_pixel_data(img_frame.pixels, keep_aspect_ratio, maximum_size)
     return None
 
 
 class NeuralWrapper:
-    def __init__(self, rob, resize, scale):
+    def __init__(self, rob, keep_aspect_ratio=False, maximum_size=(320, 240)):
```

Comment:
I would prefer keeping aspect ratio so that neural network works better without distortion

Reply:
@noskill  so your suggestion is to use keep_aspect_ratio=True as default?

Reply:
yes, because i think neural nets might work worse with distortions

Reply:
@noskill  ok, done